src/Components/Browser.JS/src/Platform/Circuits/CircuitRegistry.ts

```diff
@@ -0,0 +1,245 @@
+import { attachRootComponentToElement } from '../../Rendering/Renderer';
```

Comment:
The name of this file sounds like it reflects an earlier version of the design rather than what's here now. It's not a registry, as it's not stateful at all - it's all static methods.

Maybe we could have `ComponentEntry` in a file named `ComponentEntry.ts`, and then the static functions that deal with comment discovery in some separate `PrerenderingMarkersDiscovery.ts` that has a bunch of `export function`s instead of needing any `static class`.

I also would propose considering not doing such sophisticated discovery of malformed tag scenarios. There are a lot of lines of code for us to understand and maintain forever, but possibly not any real scenarios for them. Either the comment tags are well-formed, or the scenario is completely unsupported and fatal anyway. I'd be more confident if we ignore things that aren't well-formed - it's unclear to me what we gain from having the discriminated union types or any code path that leads to `reportInvalidPair` without just throwing at the exact point where it discovered a problem.

Reply:
Updated to throw on malformed comment pairs, and cleaned up the abstractions a bit. It started being the place to dump all the stuff that I didn't  want to put on Boot.Server.js